### PR TITLE
Extensions: WPJM - Add form reducer under state.ui subtree

### PR DESCRIPTION
**Pull request by donnapep**
The WPJM Redux Form reducer is currently mounted at `state.extensions.wpJobManager.form`. As more forms are added to more extensions, they will each place them at a different subtree location (e.g. `state.extensions.zoninator.form`). This is problematic because it means that each action that Redux Form dispatches will needlessly pass through each of these reducers in turn.

To avoid the above situation from compounding as more forms in more extensions are added, this PR moves the form reducer to `state.ui`.

Please see the comments in p4TIVU-7nc-p2c-p2 for additional context, and thanks to @jsnajdr for noticing this!

## Testing

Confirm that both the WPJM _Job Listings_ and _Job Submission_ tabs fetch and save data using [this branch](https://github.com/Automattic/wp-job-manager/tree/features/rest-api).

cc: @ice9js

### CODE REVIEW

**Comment by donnapep** [comment]
@ockham Ideally extensions should not store data outside of their designated Redux subtree, but we may need to make an exception in this case.

**Comment by donnapep** [comment]
As an alternative, we could mount the form reducer for all extensions under `state.extensions.form`, and when Calypso adds forms they could be under `state.ui.form`. This would mean that there would be still be two reducers processing all Redux Form actions, but at least when more extensions add forms the problem doesn't get bigger. And this way extensions can still keep their state in the `extensions` subtree. Thoughts?

**Comment by jsnajdr** [comment]
I like this PR a lot! It will enable us to use Redux Form at other places in Calypso that are not extensions.

When using Redux Form, extensions don't directly store any data in the Redux state tree. They do it only indirectly by using the services provided by Redux Form. And where Redux Form actually stores the data is an implementation detail that's almost completely invisible to the library user. When I think about it this way, i.e., Redux Form being just one of services provided by Calypso Framework, having the form data outside the `extensions` subtree doesn't look like a big problem to me.

To avoid naming conflict, there should be a convention that all forms registered by an extension should have `extension.extensionName.` prefix in their form ID:
```js
reduxForm( {
  form: 'extensions.wpJobManager.jobListings',
  ...
} )
```
This, I think, is the right place to worry about giving good names and prefixes to things, not the reducer.

There's one more tradeoff that I think is worth considering: what about mounting the Redux Form reducer directly at the root path of the state tree (`state.form`) instead of `state.ui.form`? While `state.ui.form` gives us a cleaner and better organized state tree, it comes with a cost: every single use of the `reduxForm` HOC and the `formValueSelector` needs to have an extra parameter with a `getFormState` selector:
```js
state => state.ui.form
```
If you forget to specify it, the form data won't be found and `ReferenceErrors` will be thrown. On the other hand, the default `state.form` works beautifully out of the box without having to think about it all the time.

File: client/extensions/wp-job-manager/components/settings/job-listings/index.jsx [hunk 1]
```diff
@@ -274,7 +274,7 @@ class JobListings extends Component {
 
 const connectComponent = connect(
 	( state ) => {
-		const selector = formValueSelector( 'jobListings', () => state.extensions.wpJobManager.form );
+		const selector = formValueSelector( 'wpJobManager.jobListings', () => state.ui.form );
```
Comment (jsnajdr) [inline client/extensions/wp-job-manager/components/settings/job-listings/index.jsx]: A little performance improvement can be done here: move the definition of the selector outside the`connect` call -- create it once when initializing the module and avoid recreating it on every state update:
```js
const selector = formValueSelector( 'wpJobManager.jobListings', state => state.ui.form );
const connectComponent = connect( state => ( {
  perPage: selector( state, 'listings.perPage' )
} ) );
```

**Comment by donnapep** [comment]
> There's one more tradeoff that I think is worth considering: what about mounting the Redux Form reducer directly at the root path of the state tree (state.form) instead of state.ui.form? While state.ui.form gives us a cleaner and better organized state tree, it comes with a cost: every single use of the reduxForm HOC and the formValueSelector needs to have an extra parameter with a getFormState selector.

@ockham Are you good with this? I agree that it would make development easier to be able to forego this extra bit of configuration code.

**Comment by ockham** [comment]
Hmm. Having it below `state.ui` would be nice in terms of how we organize our data, but if the other option allows us to drop that `state => state.ui.form`, let's go for that.

**Comment by donnapep** [comment]
@jsnajdr Made some adjustments in https://github.com/Automattic/wp-calypso/pull/16305/commits/e72f9c35c61d691aadf1b4c046c33f175425c3ad. I think we're good here now. Care to take a final review? Cheers.

**Pull request merged**